### ISSUE TITLE
vmware_guest not setting ip of Centos7 template.

### ISSUE THREAD

**Issue by natemccallum** [ansible/ansible#46175]
##### SUMMARY

Trying to deploy Centos7 template and customize the ip settings. The vm is created but there is an error with the customization. I can't tell from the log what the problem is. I've tried the suggestions for Linux.  Perl is installed and so is open-vm-tools.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

vmware_guest

##### ANSIBLE VERSION

```
ansible 2.6.4
  config file = None
  configured module search path = [u'/Users/xxxxxx/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]

```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

vSphere 6.5. Guest is Centos 7.

##### STEPS TO REPRODUCE

```yaml

```

##### EXPECTED RESULTS

VM created with IP set

##### ACTUAL RESULTS

```
changed: [localhost] => {
    "changed": true,
    "instance": {
        "annotation": "",
        "current_snapshot": null,
        "customvalues": {},
        "guest_consolidation_needed": false,
        "guest_question": null,
        "guest_tools_status": "guestToolsRunning",
        "guest_tools_version": "10282",
        "hw_cores_per_socket": 1,
        "hw_datastores": [
            "datastore1 (1)"
        ],
        "hw_esxi_host": "192.168.2.3",
        "hw_eth0": {
            "addresstype": "assigned",
            "ipaddresses": [
                "********03",
                "fe80::587c:6d37:8db4:47fa"
            ],
            "label": "Network adapter 1",
            "macaddress": "00:50:56:bb:e8:c0",
            "macaddress_dash": "00-50-56-bb-e8-c0",
            "portgroup_key": null,
            "portgroup_portkey": null,
            "summary": "VM Network"
        },
        "hw_files": [
            "[datastore1 (1)] mytestvm1/mytestvm1.vmx",
            "[datastore1 (1)] mytestvm1/mytestvm1.nvram",
            "[datastore1 (1)] mytestvm1/mytestvm1.vmsd",
            "[datastore1 (1)] mytestvm1/mytestvm1.vmdk"
        ],
        "hw_folder": "/Datacenter/vm/testvms",
        "hw_guest_full_name": "CentOS 7 (64-bit)",
        "hw_guest_ha_state": null,
        "hw_guest_id": "centos7_64Guest",
        "hw_interfaces": [
            "eth0"
        ],
        "hw_is_template": false,
        "hw_memtotal_mb": 4096,
        "hw_name": "mytestvm1",
        "hw_power_status": "poweredOn",
        "hw_processor_count": 2,
        "hw_product_uuid": "423b8081-3f3e-ed48-2bde-9bcff9a802aa",
        "hw_version": "vmx-13",
        "instance_uuid": "503b90cb-d4de-2bf6-3d8b-4e72c6122c17",
        "ipv4": "********03",
        "ipv6": "fe80::587c:6d37:8db4:47fa",
        "module_hw": true,
        "snapshots": []
    },
    "invocation": {
        "module_args": {
            "annotation": null,
            "cdrom": {},
            "cluster": null,
            "customization": {
                "dns_servers": [
                    "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                    "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
                ],
                "domain": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "hostname": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
            },
            "customization_spec": null,
            "customvalues": [],
            "datacenter": "Datacenter",
            "disk": [
                {
                    "datastore": "datastore1",
                    "size_gb": 50,
                    "type": "thin"
                }
            ],
            "esxi_hostname": "192.168.2.3",
            "folder": "/testvms",
            "force": false,
            "guest_id": "centos64Guest",
            "hardware": {
                "memory_mb": 4096,
                "num_cpus": 2
            },
            "hostname": "192.168.2.4",
            "is_template": false,
            "linked_clone": false,
            "name": "mytestvm1",
            "name_match": "first",
            "networks": [
                {
                    "dns_servers": [
                        "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                        "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
                    ],
                    "gateway": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                    "ip": "********10",
                    "name": "VM Network",
                    "netmask": "255.255.252.0",
                    "type": "static"
                }
            ],
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "resource_pool": null,
            "snapshot_src": null,
            "state": "poweredon",
            "state_change_timeout": 0,
            "template": "mytestvm",
            "username": "administrator@vsphere.local",
            "uuid": null,
            "validate_certs": false,
            "vapp_properties": [],
            "wait_for_ip_address": true
        }
    }


```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/vmware/vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @dav1x @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Hi @natemccallum,

Thank you for the issue, just so you are aware we have a dedicated Working Group for vmware.
You can find other people interested in this in #ansible-vmware on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by cigamit** [comment]
Is perl installed on the Centos 7 Template?

**Comment by natemccallum** [comment]
So that is actually interesting. I have installed Perl and net-tools to the
template. But when I deploy that template perl and net-tools commands do
not work. They simply return immediately with nothing.

For instance I do a perl -v after installing perl to a vm that I have made
from the template and I get the version info. Then stop the vm -> convert
to template. Then create a new vm from the new template I just created and
when the new vm comes up perl -v returns nothing. Same is happening with
net-tools commands like ifconfig and netstat.

Super odd.

**Comment by natemccallum** [comment]
Ok - I don't know what the difference of "convert to template" vs "clone vm
to template" but when I clone to template perl and nettools stays and the
customization works. Thanks

**Comment by ansibot** [comment]
cc @ckotte
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @nerzhul @pdellaert
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @lparkes
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Tomorrow9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Tomorrow9** [comment]
Hi @natemccallum, so there is no issue of your CentOS 7 customization anymore after Perl installed correctly? Thanks.

**Comment by natemccallum** [comment]
That is correct. But I also used "clone to template" in order to make the
template I use with ansible.

**Comment by Tomorrow9** [comment]
@natemccallum, so you mean "convert the VM to template", the configurations and installed apps in the OS are not in the converted template, but "clone to template", the cloned VM contains the changes? All these operations you did from the web client manually?

**Comment by natemccallum** [comment]
Correct. And yes all done manually to have the template I would call from
within the ansible playbook.

**Comment by Tomorrow9** [comment]
@natemccallum, ok, thanks for your reply. If so, I think it is not the "vmware_guest" module issue. Before "convert the VM to template" please make sure all the changes and installed apps are saved to disk. "convert" operation does not change the apps in the OS.

**Comment by ansibot** [comment]
cc @goneri
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @pgbidkar
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Akasurde** [comment]
@natemccallum Could you please confirm @Tomorrow9's suggestion ? Thanks,

needs_info

**Comment by natemccallum** [comment]
How do I confirm all apps are installed and saved to disk? I have installed everything and am using the same vm to clone and convert. I install my OS, install all the applications. Then do a clone which keeps the original vm and creates a template that works. Then, from the same vm I cloned from, do a convert to template and the applications are not there. I do agree this is not a vmware_guest module issue. Closing. Thanks
